### PR TITLE
feat: make executor timeout configurable (#458)

### PR DESCRIPTION
**Pull request by weibaohui**
合并 PR #458 到 main

### CODE REVIEW

**Comment by weibaohui** [comment]
## Code Review: feat: make executor timeout configurable (#458)

### Overview

This PR makes the executor timeout user-configurable and adds a UI toggle to disable it entirely. The change touches5 files (+103/-17):
- `backend/src/config.rs` — centralizes the default in `DEFAULT_EXECUTION_TIMEOUT_SECS`
- `backend/src/executor_service.rs` — supports `execution_timeout_secs ==0` meaning "no limit"
- `backend/src/handlers/config.rs` — validation accepts0 or `>=60`
- `frontend/src/components/SettingsPage.tsx` & `frontend/src/components/settings/RuntimePanel.tsx` — new Switch + InputNumber control

The implementation is correct, tests are added, and the flow `Config -> handler -> executor_service` is consistent.

### Strengths

- **Validation is well-scoped**: extracting `validate_execution_timeout_secs` with dedicated unit tests is clean, and the three boundary cases (0,59,60) are good.
- **Default centralization**: introducing `DEFAULT_EXECUTION_TIMEOUT_SECS` in `backend/src/config.rs` removes a magic number and is referenced in the `Default` impl.
- **Branching in `tokio::select!` with `if timeout_enabled`** correctly disables the timeout arm when `execution_timeout_secs ==0`, so truly unlimited execution is possible.
- **Pin-then-mut (`tokio::pin!(timeout_sleep)`)** is the right idiom for re-using a `Sleep` future across `select!` branches — the `&mut timeout_sleep` branch consumes it correctly.
- **Frontend UX** is reasonable: `lastEnabledExecutionTimeoutSecsRef` preserves the last non-zero value across toggle-off → toggle-on cycles.

### Issues & suggestions

####1. Behavior change: frontend default fallback silently went1800 →3600

In `frontend/src/components/SettingsPage.tsx` the fallback changed from `1800` to `DEFAULT_EXECUTION_TIMEOUT_SECS` (3600). If a user has `execution_timeout_secs` undefined on the form and saves, the saved value is now3600 instead of the previous1800.

This is probably intentional given the backend also defaulted to3600, but worth flagging as a behavior change in the PR description / changelog. Consider mentioning it explicitly.

####2. `DEFAULT_EXECUTION_TIMEOUT_SECS` duplicated in three places

The constant `const DEFAULT_EXECUTION_TIMEOUT_SECS =3600` is now defined in:
- `backend/src/config.rs`
- `frontend/src/components/SettingsPage.tsx`
- `frontend/src/components/settings/RuntimePanel.tsx`

Suggested fix:
- Move the constant to a single shared module, e.g. `frontend/src/constants/config.ts`, and import it where needed.
- Or have the frontend read the default from `get_config()` (already exposed) and drop the literal entirely from the frontend. This guarantees frontend and backend can never drift.

####3. `tokio::time::sleep` future is allocated even when disabled

`backend/src/executor_service.rs:709-713`

```rust
let timeout_duration = std::time::Duration::from_secs(execution_timeout_secs.max(1));
let timeout_enabled = execution_timeout_secs >0;
let timeout_mins = execution_timeout_secs /60;
let timeout_sleep = tokio::time::sleep(timeout_duration);
tokio::pin!(timeout_sleep);
```

When disabled, `timeout_sleep` still allocates a1-second `Sleep` future (because of `.max(1)`) that will never be polled in `select!` thanks to `if timeout_enabled`. Functionally correct, but slightly wasteful and confusing for readers. Consider:

```rust
let timeout_enabled = execution_timeout_secs >0;
let timeout_mins = execution_timeout_secs /60;
```

…then inline the sleep inside the (now guarded) arm, or only build the sleep when enabled. This removes the magic `.max(1)` which only exists to dodge a0-duration sleep.

####4. Mixed-language error message

`backend/src/handlers/config.rs`

Function doc is in Chinese, error message is in English:
```rust
return Err(AppError::BadRequest(
 "execution_timeout_secs must be0 or at least60".to_string(),
));
```

Other validation errors in this file use English, so the message itself is fine — but if the project aims for user-facing localization, the doc and the message should at least be aligned. Low priority.

####5. Brief UI flicker on first paint

`frontend/src/components/settings/RuntimePanel.tsx:23-24`

On first render, before `configForm` is hydrated, `executionTimeoutSecs` is initialized to `DEFAULT_EXECUTION_TIMEOUT_SECS` (3600, enabled). Once the config loads the `useEffect` syncs to the real value. If a user had previously set `execution_timeout_secs ==0`, the Switch will briefly read "开启" before flipping to "关闭". Cosmetic only.

A cleaner approach: read the value from `currentConfig` (already in scope) on first render instead of from `configForm.getFieldValue`.

####6. `Math.max(1, Math.round(...))` can mask the actual seconds

If the form ever holds a value between60 and119 seconds, the user sees `1 minute` and saves60 seconds. The backend minimum is60, so this case is currently unreachable from valid input, but the layer of rounding makes the true seconds value invisible. Consider keeping the underlying seconds precise when an InputNumber step supports it, or displaying the underlying value somewhere.

####7. Tests: backend ✅, frontend ❌

Backend has new unit tests for both `Default::default()` and the validator (3 boundary cases). The frontend logic in `RuntimePanel.tsx` (Switch toggle, ref preservation, hydration) has no automated test. Per `AGENTS.md` ("修改前端 UI 后，必须使用 Playwright 进行自动化验证"), please add at least a smoke test confirming:
- Switch toggles form value between `0` and last-enabled seconds.
- InputNumber is disabled when Switch is off.
- Form hydration sets the Switch state correctly on initial load.

### Security & Performance

- No new attack surface: validation is tighter than before.
- No new I/O or sync primitives.
- Memory: negligible — one extra `Sleep` future per execution when disabled.

### Verdict

**Approve with minor suggestions.** The core change is sound and well-tested on the backend. The two highest-leverage cleanups are: (a) consolidating the default constant to a single source, and (b) removing the `.max(1)` workaround in `executor_service.rs` now that `timeout_enabled` gates the branch. The Playwright coverage is the only gap I'd want addressed before merging into `main`.

File: backend/src/executor_service.rs [hunk 1]
```diff
@@ -706,6 +706,12 @@ pub async fn run_todo_execution(request: RunTodoExecutionRequest) -> ExecutionRe
             }
         });
 
+        let timeout_duration = std::time::Duration::from_secs(execution_timeout_secs.max(1));
```
Comment (weibaohui) [inline backend/src/executor_service.rs:709]: `timeout_sleep` is always created (1-second future due to `.max(1)`) even when `timeout_enabled == false`. The `.max(1)` was a workaround for the old code where the sleep was unconditional. Since the branch is now gated by `if timeout_enabled`, drop the `.max(1)` and either inline the sleep inside the arm or build it only when enabled — keeps the intent obvious.

File: frontend/src/components/SettingsPage.tsx [hunk 2]
```diff
@@ -122,7 +126,7 @@ export function SettingsPage({ onBack }: SettingsPageProps) {
         ...currentConfig,
         ...(values as Partial<Config>),
         max_concurrent_todos: configForm.getFieldValue('max_concurrent_todos') ?? currentConfig.max_concurrent_todos ?? 3,
-        execution_timeout_secs: configForm.getFieldValue('execution_timeout_secs') ?? currentConfig.execution_timeout_secs ?? 1800,
+        execution_timeout_secs: configForm.getFieldValue('execution_timeout_secs') ?? currentConfig.execution_timeout_secs ?? DEFAULT_EXECUTION_TIMEOUT_SECS,
```
Comment (weibaohui) [inline frontend/src/components/SettingsPage.tsx:129]: Silent behavior change: the fallback for `execution_timeout_secs` was `1800` (30 min) before this PR and is now `DEFAULT_EXECUTION_TIMEOUT_SECS` (3600,60 min). If the form value is undefined at save time, users previously saved30 min and will now save60 min. Worth calling out in the PR description / changelog.

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 3]
```diff
@@ -1,10 +1,13 @@
-import { useState, useEffect } from 'react';
-import { Card, InputNumber, Tooltip, Button, Popconfirm, Table, Empty, message } from 'antd';
+import { useState, useEffect, useRef } from 'react';
+import { Card, InputNumber, Tooltip, Button, Popconfirm, Table, Empty, Switch, message } from 'antd';
 import { InfoCircleOutlined, SaveOutlined, StopOutlined, ReloadOutlined } from '@ant-design/icons';
 import { useApp } from '@/hooks/useApp';
 import * as db from '@/utils/database';
 import type { ExecutionRecord } from '@/types';
 
+const DEFAULT_EXECUTION_TIMEOUT_SECS = 3600;
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:8]: `DEFAULT_EXECUTION_TIMEOUT_SECS =3600` is now defined in three places (config.rs, SettingsPage.tsx, here). If one changes without the others, UI and backend can drift. Consider a single `frontend/src/constants/config.ts`, or read the default from `get_config()` so the backend is the single source of truth.

File: backend/src/executor_service.rs [hunk 4]
```diff
@@ -706,6 +706,21 @@ pub async fn run_todo_execution(request: RunTodoExecutionRequest) -> ExecutionRe
...
```
Comment (weibaohui) [inline backend/src/executor_service.rs:709]: **哨兵值 `max(1)` 建议加注释。** 配合下方 `if timeout_enabled` 门禁，超时禁用时该 sleep 的真实长度其实无关紧要；用 `max(1)` 只是为了避免 `tokio::time::sleep(Duration::ZERO)` 立刻 ready 把 `select!` 抢空。读代码的人很难一眼看出这个 1 的意义，建议补一行注释说明这个设计。

更稳的写法是：

```rust
let timeout_duration = if timeout_enabled {
    Duration::from_secs(execution_timeout_secs)
} else {
    // 关闭超时：用一个几乎不会到的时长，select 真正靠 `if timeout_enabled` 门禁
    Duration::from_secs(u64::MAX)
};
```

—— 这样意图更明确、行为也更可预测（即使将来误删 `if` 门禁，进程也得跑 ~5.8 亿年才会被这条分支命中）。

File: backend/src/executor_service.rs [hunk 5]
```diff
@@ -706,6 +706,21 @@ pub async fn run_todo_execution(request: RunTodoExecutionRequest) -> ExecutionRe
...
+        let timeout_enabled = execution_timeout_secs > 0;
+        // 精确格式化超时提示：整数分钟直接显示"X 分钟"，非整数显示"X 分 Y 秒"
+        let timeout_str = {
```
Comment (weibaohui) [inline backend/src/executor_service.rs:712]: **`timeout_str` 即使在「超时禁用」路径也会被求值，是死代码 + 一次无意义的字符串拼接。** 建议把这段格式化挪到 `if timeout_enabled` 分支内（或用 `let timeout_str = timeout_enabled.then(|| { ... });`）。

File: backend/src/executor_service.rs [hunk 6]
```diff
@@ -752,10 +767,10 @@ pub async fn run_todo_execution(request: RunTodoExecutionRequest) -> ExecutionRe
                 task_manager_spawn.remove(&task_id).await;
                 return;
             }
-            _ = tokio::time::sleep(std::time::Duration::from_secs(execution_timeout_secs)) => {
+            _ = &mut timeout_sleep, if timeout_enabled => {
```
Comment (weibaohui) [inline backend/src/executor_service.rs:770]: **对应 `Duration::from_secs(... .max(1))` 的 `if timeout_enabled` 门禁本身没问题。** 提一下是因为：如果将来有人把 `timeout_duration` 改成上面建议里的 `Duration::from_secs(u64::MAX)`，需要同步评估 1) 是否会被 `biased;` 之后的子进程退出分支抢走（应该会，OK），2) 关闭超时路径下该 future 仍在 tokio 内挂起直到 select 退出——没问题，select 结束会被 drop。

File: backend/src/handlers/config.rs [hunk 7]
```diff
@@ -4,6 +4,16 @@ use crate::config::Config;
 use crate::handlers::{ApiJson, AppError, AppState};
 use crate::models::{ApiResponse, UpdateConfigRequest};
 
+/// 校验执行超时配置，允许 0 表示不限制执行时长，其余值至少为 60 秒。
```
Comment (weibaohui) [inline backend/src/handlers/config.rs:7]: **抽函数 + 单元测试好评。** 建议补一个 `execution_timeout_secs == u64::MAX` 的边界用例（虽然 3600 秒 / 1440 分钟的上限让这个值实际不会到达），以及「等于 60」已经测了，差一个「`1`、`59`、`60`、`61`」的小范围参数化测试更稳。

另外错误消息是英文，doc 注释是中文；这种「日志/校验英文、注释中文」的混排对国内项目 OK，但建议在模块顶部加一句说明以保持一致风格。

**Review by weibaohui — COMMENTED** [review]
## Code Review · PR #461 (merge-pr-458)

**范围**: 合并 PR #458（feat: make executor timeout configurable）+ 评审反馈修复（20ffc50）。改动 5 个文件、+126/-17 行。

**整体评价**: 实现思路清晰，关键校验已抽成函数、补充了单元测试；前后端对 `0 = 不限制` 的语义已对齐。下面的建议主要是健壮性、可维护性与文档同步问题，按优先级排列。

---

### 建议处理

**1. 后端：`executor_service.rs:709` 的 `max(1)` 哨兵值缺少注释，且当 `execution_timeout_secs == 0` 时 `timeout_str` 仍是死代码。**

```rust
let timeout_duration = std::time::Duration::from_secs(execution_timeout_secs.max(1));
let timeout_enabled = execution_timeout_secs > 0;
let timeout_str = { ... }; // 即使 timeout_enabled == false 也会被求值
```

`max(1)` 是为了让 `tokio::time::sleep(0)` 不会立刻 ready 把 select 抢空，但 `if timeout_enabled` 已经把这条分支门禁掉了，所以 sleep 的实际长度无关紧要。建议：
- 给这一行加注释说明「max(1) 仅为防止 0 长度 sleep 立即 ready，配合下方 `if timeout_enabled` 门禁使用」；
- 把 `timeout_str` 的格式化挪到 `if timeout_enabled` 分支内（或用 `bool` 守卫），避免禁用超时时的字符串拼接。

**2. 前端：`RuntimePanel.tsx:59-64` 的 useEffect 依赖 `configForm` 实际不会重新触发。**

```tsx
useEffect(() => {
  const formValue = configForm.getFieldValue('execution_timeout_secs');
  if (formValue !== undefined && formValue !== executionTimeoutSecs) {
    setExecutionTimeoutSecs(formValue);
  }
}, [configForm]);
```

`configForm` 是 antd 的 `FormInstance`，引用通常在父组件用 `useForm()` 一次创建后保持稳定，依赖数组里只有它，effect 实际上只在挂载时跑一次。注释提到「外部加载配置重置表单时」同步状态，但这种场景下 effect 并不会重新执行。

建议改用 antd 自带的订阅 hook：

```tsx
const formValue = Form.useWatch('execution_timeout_secs', configForm);
useEffect(() => {
  if (formValue !== undefined && formValue !== executionTimeoutSecs) {
    setExecutionTimeoutSecs(formValue);
  }
}, [formValue, executionTimeoutSecs]);
```

这样配置异步加载完成后状态能正确同步，也不会被 `configForm` 引用稳定性所限制。

**3. 前端：`RuntimePanel.tsx:33` 的 `lastEnabledExecutionTimeoutSecsRef` 初始值固定为 3600。**

```tsx
const lastEnabledExecutionTimeoutSecsRef = useRef<number>(DEFAULT_EXECUTION_TIMEOUT_SECS);
```

如果用户加载到的配置本来就是 `execution_timeout_secs = 0`（超时关闭），第一次点「开启」时值会跳到 3600 而非恢复原配置。建议用懒初始化从表单读一次：

```tsx
const lastEnabledExecutionTimeoutSecsRef = useRef<number>(
  configForm.getFieldValue('execution_timeout_secs') || DEFAULT_EXECUTION_TIMEOUT_SECS
);
```

**4. 文档：`docs/user-guide/settings/system-settings.md` 未同步「0 = 不限制」语义。**

`execution_timeout_secs` 现在新增了「`0` 表示不限制执行时长」的能力，但用户设置页和文档里都没说明。建议在 system-settings.md 第 28 行的表格中加注：

> `execution_timeout_secs = 0` 表示不限制执行时长；其他合法值需 ≥ 60 秒。

runtime-management.md 的相关小节也建议同步。

**5. 前端：`RuntimePanel.tsx:136-141` 的 `if (v)` 会静默吞掉清空操作。**

```tsx
onChange={(v) => {
  if (v) {
    setExecutionTimeoutSecs(v * 60);
    configForm.setFieldsValue({ execution_timeout_secs: v * 60 });
  }
}}
```

如果用户清空输入框，`v` 是 `null`，整个分支不执行；本地 state 和表单值都会保持原值，UX 上会让人困惑。建议至少让 `setExecutionTimeoutSecs` 跟着 `setFieldsValue` 一起更新（或显式把状态重置为 0，并交给后端校验）。

**6. 前后端精度往返：`Math.round(executionTimeoutSecs / 60)` 在配置非整分钟时会丢精度。**

- 后端最小粒度是 1 秒；
- 前端 `InputNumber` 步长为 1 分钟，写入时 `v * 60`；
- 读取时 `Math.round(... / 60)` 假设底层是整分钟。

如果用户手动改 YAML 或 API 把 `execution_timeout_secs` 写成 90，前端会显示 `2` 分钟、回写时变成 120。建议读取时也用「分钟数 + 隐藏秒数」的形式（InputNumber 步长=1 秒会更精确）；或者文档化「UI 以分钟为单位、底层以秒为单位」。

**7. 缺测试：`executor_service.rs` 新逻辑无单元测试。**

新增的 `timeout_enabled` 分支、`timeout_str` 格式化、「`0` = 不超时」路径目前都未覆盖。可考虑：
- 抽出 `format_timeout_str(secs: u64) -> String` 单独测；
- 给 `run_todo_execution` 加一个注入 `execution_timeout_secs = 0` 的集成测试，断言分支不被触发。

---

### 可选改进（小问题）

- `RuntimePanel.tsx:41` 注释 `// 中文：加载运行中任务失败` 和英文 `console.error` 内容是同一信息的两种语言，长期看易漂移；建议保留英文日志监控语义、注释只写「保留英文以便 ELK 之类系统聚合」之类。
- `configForm: any` 仍是 `any`，对前端来说可以替换为 `FormInstance<ConfigFormValues>` 之类的具体类型，新逻辑（`useWatch`、`setFieldsValue`）能从类型检查受益。
- `SettingsPage.tsx:129` 的回退链 `?? currentConfig.execution_timeout_secs ?? DEFAULT_EXECUTION_TIMEOUT_SECS` 把硬编码 1800 改成 3600 是合理的，但更稳的写法是只信任 `currentConfig`，因为「表单值在保存前不可信」——若表单值是 `undefined` 且 `currentConfig` 也是 `undefined`（旧配置）才用默认值。

---

合并结论：**Approve with comments**。建议把 #1（max(1) 注释 + timeout_str 死代码）和 #2（`useEffect` 不重跑）改掉再合入；其余可后续跟进。

File: backend/src/config.rs [hunk 8]
```diff
@@ -322,6 +324,15 @@ mod tests {
         assert_eq!(cfg.server_url(), "http://localhost:9090");
     }
 
+    #[test]
```
Comment (weibaohui) [inline backend/src/config.rs:327]: **新增的 `test_default_execution_timeout_round_trip` 改进得不错。** 顺便提一下：原 `test_round_trip`（`{ port: 1234, ..Default::default() }`）和这个新用例覆盖了不同路径，可以再补一个「`execution_timeout_secs = 0` → 序列化 → 反序列化 → 仍为 0」明确锁定禁用超时的往返语义，避免以后 `#[serde(default)]` 之类被误改成回退到 3600。

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 9]
```diff
@@ -16,13 +19,26 @@ export function RuntimePanel({ configForm, configSaving, handleSaveConfig, execu
   const [selectedRecordIds, setSelectedRecordIds] = useState<number[]>([]);
   const [stoppingRecords, setStoppingRecords] = useState(false);
   const [runningRecords, setRunningRecords] = useState<ExecutionRecord[]>([]);
+  // 使用懒初始化避免表单未填充时读到默认值：初始化时若表单已有值则用表单的，否则用常量默认值。
+  // 后续通过 useEffect 同步表单值变化，保持 UI 与表单状态一致。
+  const [executionTimeoutSecs, setExecutionTimeoutSecs] = useState<number>(() =>
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:24]: **懒初始化没问题，但和第 59 行 useEffect 的配合有竞态：** 挂载时如果 `configForm` 还没拿到值（例如配置是异步加载的），`executionTimeoutSecs` 会先用 3600 兜底；第 59 行那个 effect 依赖只有 `configForm`，而 antd 的 `FormInstance` 引用在父组件用 `useForm()` 创建后基本稳定——所以这个 effect 实际上只在挂载时跑一次，不会再响应后续的 `setFieldsValue`/`resetFields`。

建议改用 antd 的订阅 hook：

```tsx
import { Form } from 'antd';
const watched = Form.useWatch('execution_timeout_secs', configForm);
useEffect(() => {
  if (watched !== undefined && watched !== executionTimeoutSecs) {
    setExecutionTimeoutSecs(watched);
  }
}, [watched, executionTimeoutSecs]);
```

这样配置异步加载、或者父组件后续 reset 表单时，本地 state 都能正确跟随。

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 10]
```diff
@@ -16,13 +19,26 @@ export function RuntimePanel({ configForm, configSaving, handleSaveConfig, execu
...
+    configForm.getFieldValue('execution_timeout_secs') ?? DEFAULT_EXECUTION_TIMEOUT_SECS
+  );
+  // 0 表示禁用执行超时（与后端 handlers/config.rs 中的校验对齐），其余值至少为 60 秒
+  const executionTimeoutEnabled = executionTimeoutSecs !== 0;
+  const executionTimeoutMinutes = executionTimeoutEnabled
+    ? Math.max(1, Math.round(executionTimeoutSecs / 60))
+    : undefined;
+  // 关闭时记录当前值，重新开启时恢复；避免再次开启时跳回初始默认值 3600。
+  const lastEnabledExecutionTimeoutSecsRef = useRef<number>(DEFAULT_EXECUTION_TIMEOUT_SECS);
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:33]: **初始值固定 3600 在一种边界场景下会跳变。** 如果用户加载的配置本来就是 `execution_timeout_secs = 0`（超时关闭），第一次点「开启」时会从 0 跳到 3600，而不是恢复原配置值。建议改成懒初始化读一次表单：

```tsx
const lastEnabledExecutionTimeoutSecsRef = useRef<number>(
  configForm.getFieldValue('execution_timeout_secs') || DEFAULT_EXECUTION_TIMEOUT_SECS
);
```

注意：`||` 会把 `0` 也当成「未设置」而落到 3600；这里恰好就是我们想要的「不限制」回退到默认 3600 的语义，所以用 `||` 是对的。如果担心 `null` 之外的 falsy 值，再单独 `?? DEFAULT_EXECUTION_TIMEOUT_SECS`。

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 11]
```diff
@@ -32,6 +48,22 @@ export function RuntimePanel({ configForm, configSaving, handleSaveConfig, execu
...
 
+  // 当用户手动输入时记录最新值；若关闭超时则不更新（保留重新开启时的恢复值）。
+  useEffect(() => {
+    if (executionTimeoutEnabled) {
+      lastEnabledExecutionTimeoutSecsRef.current = executionTimeoutSecs;
+    }
+  }, [executionTimeoutEnabled, executionTimeoutSecs]);
+
+  // 监听表单字段变化（外部加载配置重置表单时），同步本地状态。
+  useEffect(() => {
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:59]: **和 #24 同一类问题：依赖 `[configForm]` 实际不会重跑。** 见上面 #24 的建议用 `Form.useWatch`。

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 12]
```diff
@@ -76,20 +117,31 @@ export function RuntimePanel({ configForm, configSaving, handleSaveConfig, execu
...
+            />
             <InputNumber
               size="small"
               min={1}
               max={1440}
               style={{ width: 80 }}
-              value={Math.round((configForm.getFieldValue('execution_timeout_secs') ?? 1800) / 60)}
+              disabled={!executionTimeoutEnabled}
+              value={executionTimeoutMinutes}
               onChange={(v) => {
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:136]: **`if (v)` 静默吞掉清空操作是个 UX 陷阱。** 用户在 InputNumber 里把值删掉，`v` 是 `null`，整个分支不执行——本地 state 和表单值都保持原值，光标离开时用户会发现「我明明删了，怎么还在」。

至少加一个 `else` 把状态显式重置；具体目标值（0 vs 后端最小值 60）可以参考后端 `validate_execution_timeout_secs`：

```tsx
onChange={(v) => {
  const secs = v ? v * 60 : 0;
  setExecutionTimeoutSecs(secs);
  configForm.setFieldsValue({ execution_timeout_secs: secs });
}}
```

这样清空 = 禁用超时，和「关闭」开关的语义一致。

File: frontend/src/components/settings/RuntimePanel.tsx [hunk 13]
```diff
@@ -76,20 +117,31 @@ export function RuntimePanel({ configForm, configSaving, handleSaveConfig, execu
...
                 if (v) {
+                  setExecutionTimeoutSecs(v * 60);
                   configForm.setFieldsValue({ execution_timeout_secs: v * 60 });
                 }
               }}
             />
-            <Tooltip title="单个对话执行的最大时长，超时将自动终止">
+            <span style={{ fontSize: 13, color: 'var(--color-text-secondary)', whiteSpace: 'nowrap' }}>分钟</span>
+            <Tooltip title="单个执行任务的最大时长；关闭后不再因超时自动终止">
```
Comment (weibaohui) [inline frontend/src/components/settings/RuntimePanel.tsx:144]: **Tooltip 文案好评（关 = 不再因超时自动终止），但 UI 上没有显式提示「当前是关闭状态」。** Switch 关闭时 InputNumber 是灰的 + 空白，用户得 hover Tooltip 才知道是关的。建议：
- 在 InputNumber 旁边加一行小字「当前不限制执行时长」
- 或在 Switch 关闭时把右侧「分钟」label 替换成「无限制」

这是 UX 改进项，不是阻塞性问题。

File: frontend/src/components/SettingsPage.tsx (hunk 2 above)
Comment (weibaohui) [inline frontend/src/components/SettingsPage.tsx:129]: **回退链 `?? currentConfig.execution_timeout_secs ?? DEFAULT_EXECUTION_TIMEOUT_SECS` 把原来硬编码 1800 改到 3600 没问题。** 但更稳的语义是：

- 「表单值」在保存前不可信——只用它覆盖用户当前编辑的值；
- 「`currentConfig`」是后端给的权威值；
- 默认值只在「两者都没有」时才用。

建议明确这个意图（也可以加注释）：

```tsx
execution_timeout_secs: configForm.getFieldValue('execution_timeout_secs')
  ?? currentConfig.execution_timeout_secs
  ?? DEFAULT_EXECUTION_TIMEOUT_SECS,
```

—— 这正是当前写法，只是 `currentConfig.execution_timeout_secs` 的类型是 `u64` 不会为 `undefined`，所以最后一截 `?? DEFAULT_EXECUTION_TIMEOUT_SECS` 永远拿不到。考虑直接去掉最后一截，或者加注释说明「仅为类型安全保留」。
